scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,398 @@
+    direction is also in the graph (possibly with capacity 0).
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex v, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    tails : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``tails[e]`` is the tail vertex of ``e``.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``tails[e]`` is the head vertex of ``e``.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    rev_edge_ptr : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``rev_edge_ptr[e]`` is the edge obtained by
+        reversing ``e``. In particular, ``rev_edge_ptr[rev_edge_ptr[e]] == e``.
+    source : int
+        The source vertex.
+    sink : int
+        The sink vertex.
+
+    Returns
+    -------
+    flow : memoryview of length :math:`|E|`
+        The residual graph with respect to a maximum flow.
+
+    """
+    cdef int n_verts = edge_ptr.shape[0] - 1
+    cdef int n_edges = capacities.shape[0]
+
+    # Our result array will keep track of the flow along each edge
+    cdef int[:] flow = np.zeros(n_edges, dtype=ITYPE)
+
+    # Create a circular queue for breadth-first search. Elements are
+    # popped dequeued at index start and queued at index end.
+    cdef int[:] q = np.empty(n_verts, dtype=ITYPE)
+    cdef int start, end
+
+    # Create an array indexing predecessor edges
+    cdef int[:] pred_edge = np.empty(n_verts, dtype=ITYPE)
+
+    cdef bint path_found
+    cdef int cur, df, t, e, edge, k
+
+    # While augmenting paths from source to sink exist
+    while True:
+        for k in range(n_verts):
+            pred_edge[k] = -1
+        # Reset queue to consist only of source
+        q[0] = source
+        start = 0
+        end = 1
+        # While we have not found a path, and queue is not empty
+        path_found = False
+        while start != end and not path_found:
+            # Pop queue
+            cur = q[start]
+            start += 1
+            if start == n_verts:
+                start = 0
+            # Loop over all edges from the current vertex
+            for e in range(edge_ptr[cur], edge_ptr[cur + 1]):
+                t = heads[e]
+                if pred_edge[t] == -1 and t != source and\
+                        capacities[e] > flow[e]:
+                    pred_edge[t] = e
+                    if t == sink:
+                        path_found = True
+                        break
+                    # Push to queue
+                    q[end] = t
+                    end += 1
+                    if end == n_verts:
+                        end = 0
+        # Did we find an augmenting path?
+        if path_found:
+            df = 2147483647  # Max 32-bit signed int
```

Comment:
We may forget to tweak this is `ITYPE` ever changes, so I suggest to use
```
ITYPE_MAX = np.iinfo(ITYPE).max
```
outside the main loop, and then use `ITYPE_MAX` here